### DISCUSSION TITLE
Dynamically set values in `click.Choice` type

### DISCUSSION THREAD

**Discussion by ShawnRG** [pallets/click#2646] [Q&A]
Is it possible to dynamically set the choices on a `click.Choice` option? 
The added complexity here is that, depending on the value of an option set on a group, the value will change.
Here is a small example of what I am trying to achieve:
```python
@click.group
@click.option('--env')
def init(env):
    load_command_choices(env)
    pass

@init.command
@click.option('-c', type=click.Choice(get_command_choices()))
def some_command(c):
##
```

This works fine when you do not provide the `-c` argument explicitly, but fails when you do with the following error:
```
Usage: app init some-command [OPTIONS]
Try 'app init some-command --help' for help.

Error: Invalid value for '-c': 'test-value' is not one of .
``` 

I have tried the following:

- Make use of a callback on the command level to fill in the data: Callback is never triggered when providing the `-c` argument

Kind regards

**Comment by davidism** [comment]
You'd need to write a custom type to do this, but it should be possible, as the `ParamType` methods are passed the current context that could be used to introspect already parsed options. Note that you're also imposing ordering to the options, since you'd need the first to be processed to affect the options in the second, and we generally want to avoid a required order to options.

**Reply by ShawnRG** [reply]
Thanks for the quick answer. I will look into creating a custom type. As for the order of the arguments, in my example, the `--env` value will always come before the command's option, no?

**Reply by davidism** [reply]
Oh, in this case you're using a group, so yes.

**Comment by Connor1McKee** [comment]
Looks like this might still be actively affecting users, encountered this same thing and it looks like you would definitely have to custom type this. Tried several differently typed objects and expansion methods, including using the future annotations import with no luck for extra info.
